pkg/controller/cloud/node_gc_controller_test.go

```diff
@@ -0,0 +1,225 @@
+*/
+
+package cloud
+
+import (
+	"testing"
+	"time"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/record"
+	fakecloud "k8s.io/kubernetes/pkg/cloudprovider/providers/fake"
+	"k8s.io/kubernetes/pkg/controller/testutil"
+	schedulerapi "k8s.io/kubernetes/pkg/scheduler/api"
+
+	"github.com/golang/glog"
+	"github.com/stretchr/testify/assert"
+)
+
+// This test checks that the node is deleted when kubelet stops reporting
+// and cloud provider says node is gone
+func TestNodeDeleted(t *testing.T) {
+	pod0 := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "default",
+			Name:      "pod0",
+		},
+		Spec: v1.PodSpec{
+			NodeName: "node0",
+		},
+		Status: v1.PodStatus{
+			Conditions: []v1.PodCondition{
+				{
+					Type:   v1.PodReady,
+					Status: v1.ConditionTrue,
+				},
+			},
+		},
+	}
+
+	pod1 := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "default",
+			Name:      "pod1",
+		},
+		Spec: v1.PodSpec{
+			NodeName: "node0",
+		},
+		Status: v1.PodStatus{
+			Conditions: []v1.PodCondition{
+				{
+					Type:   v1.PodReady,
+					Status: v1.ConditionTrue,
+				},
+			},
+		},
+	}
+
+	fnh := &testutil.FakeNodeHandler{
+		Existing: []*v1.Node{
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:              "node0",
+					CreationTimestamp: metav1.Date(2012, 1, 1, 0, 0, 0, 0, time.UTC),
+				},
+				Status: v1.NodeStatus{
+					Conditions: []v1.NodeCondition{
+						{
+							Type:               v1.NodeReady,
+							Status:             v1.ConditionUnknown,
+							LastHeartbeatTime:  metav1.Date(2015, 1, 1, 12, 0, 0, 0, time.UTC),
+							LastTransitionTime: metav1.Date(2015, 1, 1, 12, 0, 0, 0, time.UTC),
+						},
+					},
+				},
+			},
+		},
+		Clientset:      fake.NewSimpleClientset(&v1.PodList{Items: []v1.Pod{*pod0, *pod1}}),
+		DeleteWaitChan: make(chan struct{}),
+	}
+
+	eventBroadcaster := record.NewBroadcaster()
+	cloudNodeDelController := &CloudNodeGCController{
+		kubeClient: fnh,
+		cloud: &fakecloud.FakeCloud{
+			ExistsByProviderID: false,
+			Err:                nil,
+		},
+		recorder:          eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "cloud-node-gc-controller"}),
+		nodeMonitorPeriod: 1 * time.Second,
+	}
+	eventBroadcaster.StartLogging(glog.Infof)
+
+	cloudNodeDelController.Run(wait.NeverStop)
+
+	select {
+	case <-fnh.DeleteWaitChan:
+	case <-time.After(wait.ForeverTestTimeout):
+		t.Errorf("Timed out waiting %v for node to be deleted", wait.ForeverTestTimeout)
+	}
+
+	assert.Equal(t, 1, len(fnh.DeletedNodes), "Node was not deleted")
+	assert.Equal(t, "node0", fnh.DeletedNodes[0].Name, "Node was not deleted")
+}
+
+func TestNodeShutdown(t *testing.T) {
+
+	testCases := []struct {
+		testName           string
+		node               *v1.Node
+		existsByProviderID bool
+		shutdown           bool
```

Comment:
Sadly a lot of the cloud providers have not implemented this call yet and so rather than returning true/false they return a not implemented error. Can we have a test for that case? (For GCE for instance it would return error for shutdown but would correctly return true/false for existsByProviderID)

Reply:
done